// @stylable/core/dist/src/stylable-resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stylable_value_parsers_1 = require("./stylable-value-parsers");
/* tslint:disable:max-line-length */
exports.resolverWarnings = {
    UNKNOWN_IMPORTED_FILE: function (path) { return "cannot resolve imported file: \"" + path + "\""; },
    UNKNOWN_IMPORTED_SYMBOL: function (name, path) { return "cannot resolve imported symbol \"" + name + "\" in stylesheet \"" + path + "\""; }
};
var StylableResolver = /** @class */ (function () {
    function StylableResolver(fileProcessor, requireModule) {
        this.fileProcessor = fileProcessor;
        this.requireModule = requireModule;
    }
    StylableResolver.prototype.resolveImported = function (imported, name) {
        var context = imported.context, from = imported.from;
        var symbol;
        if (from.match(/\.css$/)) {
            var meta = void 0;
            try {
                meta = this.fileProcessor.process(from, false, context);
                symbol = !name ? meta.mappedSymbols[meta.root] : meta.mappedSymbols[name];
            }
            catch (e) {
                return null;
            }
            return { _kind: 'css', symbol: symbol, meta: meta };
        }
        else {
            var _module = void 0;
            try {
                _module = this.requireModule(from);
            }
            catch (_a) {
                return null;
            }
            symbol = !name ?
                _module.default || _module :
                _module[name];
            return { _kind: 'js', symbol: symbol, meta: null };
        }
    };
    StylableResolver.prototype.resolveImport = function (importSymbol) {
        var name = importSymbol.type === 'named' ?
            importSymbol.name :
            '';
        return this.resolveImported(importSymbol.import, name);
    };
    StylableResolver.prototype.resolve = function (maybeImport) {
        if (!maybeImport || maybeImport._kind !== 'import') {
            if (maybeImport && maybeImport._kind !== 'var') {
                if (maybeImport.alias && !maybeImport[stylable_value_parsers_1.valueMapping.extends]) {
                    maybeImport = maybeImport.alias;
                }
                else if (maybeImport[stylable_value_parsers_1.valueMapping.extends]) {
                    maybeImport = maybeImport[stylable_value_parsers_1.valueMapping.extends];
                }
                else {
                    return null;
                }
            }
            else {
                return null;
            }
        }
        if (!maybeImport || maybeImport._kind !== 'import') {
            return null;
        }
        return this.resolveImport(maybeImport);
    };
    StylableResolver.prototype.deepResolve = function (maybeImport, path) {
        if (path === void 0) { path = []; }
        var resolved = this.resolve(maybeImport);
        while (resolved &&
            resolved._kind === 'css' &&
            resolved.symbol &&
            resolved.symbol._kind === 'import') {
            resolved = this.resolve(resolved.symbol);
        }
        if (resolved &&
            resolved.symbol &&
            resolved.meta &&
            (resolved.symbol._kind === 'class' || resolved.symbol._kind === 'element') &&
            resolved.symbol.alias &&
            !resolved.symbol[stylable_value_parsers_1.valueMapping.extends]) {
            if (path.indexOf(resolved.symbol) !== -1) {
                return { _kind: 'css', symbol: resolved.symbol, meta: resolved.meta };
            }
            path.push(resolved.symbol);
            return this.deepResolve(resolved.symbol.alias, path);
        }
        return resolved;
    };
    StylableResolver.prototype.resolveSymbolOrigin = function (symbol, meta, path) {
        if (path === void 0) { path = []; }
        if (!symbol || !meta) {
            return null;
        }
        if (symbol._kind === 'element' || symbol._kind === 'class') {
            if (path.indexOf(symbol) !== -1) {
                return { meta: meta, symbol: symbol, _kind: 'css' };
            }
            path.push(symbol);
            var isAliasOnly = symbol.alias && !symbol[stylable_value_parsers_1.valueMapping.extends];
            return isAliasOnly
                ? this.resolveSymbolOrigin(symbol.alias, meta, path)
                : { meta: meta, symbol: symbol, _kind: 'css' };
        }
        else if (symbol._kind === 'import') {
            var resolved = this.resolveImport(symbol);
            if (resolved && resolved.symbol && resolved._kind === 'css') {
                return this.resolveSymbolOrigin(resolved.symbol, resolved.meta, path);
            }
            else {
                return null;
            }
        }
        return null;
    };
    StylableResolver.prototype.resolveClass = function (meta, symbol) {
        return this.resolveName(meta, symbol, false);
    };
    StylableResolver.prototype.resolveName = function (meta, symbol, isElement) {
        var type = isElement ? 'element' : 'class';
        var finalSymbol;
        var finalMeta;
        if (symbol._kind === type) {
            finalSymbol = symbol;
            finalMeta = meta;
        }
        else if (symbol._kind === 'import') {
            var resolved = this.deepResolve(symbol);
            if (resolved && resolved._kind === 'css' && resolved.symbol) {
                if (resolved.symbol._kind === 'class' || resolved.symbol._kind === 'element') {
                    finalSymbol = resolved.symbol;
                    finalMeta = resolved.meta;
                }
                else {
                    // TODO: warn
                }
            }
            else {
                // TODO: warn
            }
        }
        else {
            // TODO: warn
        }
        if (finalMeta && finalSymbol) {
            return {
                _kind: 'css',
                symbol: finalSymbol,
                meta: finalMeta
            };
        }
        else {
            return null;
        }
    };
    StylableResolver.prototype.resolveElement = function (meta, symbol) {
        return this.resolveName(meta, symbol, true);
    };
    StylableResolver.prototype.resolveExtends = function (meta, className, isElement, transformer) {
        if (isElement === void 0) { isElement = false; }
        var bucket = isElement ? meta.elements : meta.classes;
        var type = isElement ? 'element' : 'class';
        var customSelector = isElement ? null : meta.customSelectors[':--' + className];
        if (!bucket[className] && !customSelector) {
            return [];
        }
        if (customSelector && transformer) {
            var parsed = transformer.resolveSelectorElements(meta, customSelector);
            if (parsed.length === 1) {
                return parsed[0][parsed[0].length - 1].resolved;
            }
            else {
                return [];
            }
        }
        var extendPath = [];
        var resolvedClass = this.resolveName(meta, bucket[className], isElement);
        if (resolvedClass &&
            resolvedClass._kind === 'css' &&
            resolvedClass.symbol &&
            resolvedClass.symbol._kind === type) {
            var current = resolvedClass;
            var extend = resolvedClass.symbol[stylable_value_parsers_1.valueMapping.extends] || resolvedClass.symbol.alias;
            while (current) {
                extendPath.push(current);
                if (!extend) {
                    break;
                }
                var res = this.resolve(extend);
                if (res &&
                    res._kind === 'css' &&
                    res.symbol &&
                    (res.symbol._kind === 'element' || res.symbol._kind === 'class')) {
                    current = res;
                    extend = res.symbol[stylable_value_parsers_1.valueMapping.extends] || res.symbol.alias;
                }
                else {
                    break;
                }
            }
        }
        return extendPath;
    };
    StylableResolver.prototype.validateImports = function (meta, diagnostics) {
        for (var _i = 0, _a = meta.imports; _i < _a.length; _i++) {
            var importObj = _a[_i];
            var resolvedImport = this.resolveImported(importObj, '');
            if (!resolvedImport) {
                // warn about unknown imported files
                var fromDecl = importObj.rule.nodes &&
                    importObj.rule.nodes.find(function (decl) { return decl.type === 'decl' && decl.prop === stylable_value_parsers_1.valueMapping.from; });
                if (fromDecl) {
                    diagnostics.warn(fromDecl, exports.resolverWarnings.UNKNOWN_IMPORTED_FILE(importObj.fromRelative), { word: importObj.fromRelative });
                }
            }
            else if (resolvedImport._kind === 'css') {
                // warn about unknown named imported symbols
                for (var name_1 in importObj.named) {
                    var origName = importObj.named[name_1];
                    var resolvedSymbol = this.resolveImported(importObj, origName);
                    var namedDecl = importObj.rule.nodes &&
                        importObj.rule.nodes.find(function (decl) { return decl.type === 'decl' && decl.prop === stylable_value_parsers_1.valueMapping.named; });
                    if (!resolvedSymbol.symbol && namedDecl) {
                        diagnostics.warn(namedDecl, exports.resolverWarnings.UNKNOWN_IMPORTED_SYMBOL(origName, importObj.fromRelative), { word: origName });
                    }
                }
            }
        }
    };
    return StylableResolver;
}());
exports.StylableResolver = StylableResolver;
