// yoshi-config/build/globs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scss = exports.less = exports.multipleModules = exports.singleModule = exports.testFilesWatch = exports.unitTests = exports.e2eTests = exports.specs = exports.babel = exports.dist = exports.assetsBase = exports.assetsLegacyBaseDirs = exports.baseDirs = exports.statics = void 0;
const path_1 = __importDefault(require("path"));
const distDir = 'dist';
const esModulesDist = path_1.default.join(distDir, 'es');
exports.statics = path_1.default.join(distDir, 'statics');
exports.baseDirs = [
    'app',
    'src',
    'bin',
    'test',
    '__tests__',
    'testkit',
    'stories',
    'packages',
];
exports.assetsLegacyBaseDirs = [
    'app',
    'bin',
    'test',
    'testkit',
    'stories',
];
exports.assetsBase = 'src';
exports.dist = ({ esTarget = false } = {}) => esTarget ? esModulesDist : distDir;
exports.babel = [
    ...exports.baseDirs.map(dir => path_1.default.join(dir, '**', '*.js{,x}')),
    'index.js',
];
exports.specs = exports.baseDirs.map(dir => `${dir}/**/*.+(spec|it).+(ts|js){,x}`);
exports.e2eTests = exports.baseDirs.map(dir => `${dir}/**/*.e2e.+(ts|js){,x}`);
exports.unitTests = exports.baseDirs.map(dir => `${dir}/**/*.spec.+(ts|js){,x}`);
exports.testFilesWatch = [
    ...exports.baseDirs.map(dir => path_1.default.join(dir, '**', '*.(ts|js){,x}')),
    'index.js',
];
exports.singleModule = {
    clientDist: exports.statics,
};
exports.multipleModules = {
    clientDist: distDir,
};
exports.less = [
    ...exports.baseDirs.map(dir => `${dir}/**/*.less`),
    ...exports.baseDirs.map(dir => `!${dir}/assets/**/*`),
];
exports.scss = [
    ...exports.baseDirs.map(dir => `${dir}/**/*.scss`),
    ...exports.baseDirs.map(dir => `!${dir}/assets/**/*`),
];
