// yoshi-config/build/normalize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const globs_1 = require("./globs");
exports.default = (initialConfig, pkgJson) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29;
    const { name, unpkg, dependencies = {}, peerDependencies = {}, jest = {}, } = pkgJson;
    const cdnPort = (_c = (_b = (_a = initialConfig.servers) === null || _a === void 0 ? void 0 : _a.cdn) === null || _b === void 0 ? void 0 : _b.port) !== null && _c !== void 0 ? _c : 3200;
    const cdnSsl = (_f = (_e = (_d = initialConfig.servers) === null || _d === void 0 ? void 0 : _d.cdn) === null || _e === void 0 ? void 0 : _e.ssl) !== null && _f !== void 0 ? _f : false;
    const cdnUrl = (_j = (_h = (_g = initialConfig.servers) === null || _g === void 0 ? void 0 : _g.cdn) === null || _h === void 0 ? void 0 : _h.url) !== null && _j !== void 0 ? _j : `${cdnSsl ? 'https:' : 'http:'}//localhost:${cdnPort}/`;
    const cdn = {
        port: cdnPort,
        ssl: cdnSsl,
        url: cdnUrl,
    };
    const clientProjectName = initialConfig.clientProjectName;
    const singleDir = (_m = (_l = (_k = initialConfig.servers) === null || _k === void 0 ? void 0 : _k.cdn) === null || _l === void 0 ? void 0 : _l.dir) !== null && _m !== void 0 ? _m : globs_1.singleModule.clientDist;
    const multiDir = (_q = (_p = (_o = initialConfig.servers) === null || _o === void 0 ? void 0 : _o.cdn) === null || _p === void 0 ? void 0 : _p.dir) !== null && _q !== void 0 ? _q : globs_1.multipleModules.clientDist;
    const clientFilesPath = clientProjectName
        ? `node_modules/${clientProjectName}/${multiDir}`
        : singleDir;
    const config = {
        name: name,
        unpkg,
        pkgJson,
        servers: {
            cdn,
            app: {
                port: (_t = (_s = (_r = initialConfig.servers) === null || _r === void 0 ? void 0 : _r.app) === null || _s === void 0 ? void 0 : _s.port) !== null && _t !== void 0 ? _t : (parseInt(process.env.PORT || '', 10) || 3000),
            },
        },
        specs: (_u = initialConfig.specs) !== null && _u !== void 0 ? _u : {},
        hooks: (_v = initialConfig.hooks) !== null && _v !== void 0 ? _v : {},
        hmr: (_w = initialConfig.hmr) !== null && _w !== void 0 ? _w : true,
        liveReload: (_x = initialConfig.liveReload) !== null && _x !== void 0 ? _x : true,
        exports: initialConfig.exports,
        entry: initialConfig.entry,
        splitChunks: (_y = initialConfig.splitChunks) !== null && _y !== void 0 ? _y : false,
        separateCss: (_z = initialConfig.separateCss) !== null && _z !== void 0 ? _z : true,
        cssModules: (_0 = initialConfig.cssModules) !== null && _0 !== void 0 ? _0 : true,
        tpaStyle: (_1 = initialConfig.tpaStyle) !== null && _1 !== void 0 ? _1 : false,
        enhancedTpaStyle: (_2 = initialConfig.enhancedTpaStyle) !== null && _2 !== void 0 ? _2 : false,
        separateStylableCss: (_3 = initialConfig.separateStylableCss) !== null && _3 !== void 0 ? _3 : false,
        features: (_4 = initialConfig.features) !== null && _4 !== void 0 ? _4 : {},
        externals: (_5 = initialConfig.externals) !== null && _5 !== void 0 ? _5 : [],
        transpileTests: (_6 = initialConfig.transpileTests) !== null && _6 !== void 0 ? _6 : true,
        externalUnprocessedModules: (_7 = initialConfig.externalUnprocessedModules) !== null && _7 !== void 0 ? _7 : [],
        externalizeRelativeLodash: (_9 = (_8 = initialConfig.features) === null || _8 === void 0 ? void 0 : _8.externalizeRelativeLodash) !== null && _9 !== void 0 ? _9 : false,
        petriSpecsConfig: (_10 = initialConfig.petriSpecs) !== null && _10 !== void 0 ? _10 : {},
        performanceBudget: (_11 = initialConfig.performance) !== null && _11 !== void 0 ? _11 : false,
        resolveAlias: (_12 = initialConfig.resolveAlias) !== null && _12 !== void 0 ? _12 : {},
        startUrl: (_13 = initialConfig.startUrl) !== null && _13 !== void 0 ? _13 : null,
        keepFunctionNames: (_14 = initialConfig.keepFunctionNames) !== null && _14 !== void 0 ? _14 : false,
        umdNamedDefine: (_15 = initialConfig.umdNamedDefine) !== null && _15 !== void 0 ? _15 : true,
        experimentalBuildHtml: (_16 = initialConfig.experimentalBuildHtml) !== null && _16 !== void 0 ? _16 : false,
        experimentalMonorepo: (_17 = initialConfig.experimentalMonorepo) !== null && _17 !== void 0 ? _17 : false,
        experimentalMinimalPRBuild: (_18 = initialConfig.experimentalBuildHtml) !== null && _18 !== void 0 ? _18 : false,
        experimentalRtlCss: (_19 = initialConfig.experimentalRtlCss) !== null && _19 !== void 0 ? _19 : false,
        experimentalUseAssetRelocator: (_20 = initialConfig.experimentalUseAssetRelocator) !== null && _20 !== void 0 ? _20 : false,
        yoshiServer: (_21 = initialConfig.yoshiServer) !== null && _21 !== void 0 ? _21 : false,
        projectType: (_22 = initialConfig.projectType) !== null && _22 !== void 0 ? _22 : null,
        webWorkerEntry: (_23 = initialConfig.webWorker) === null || _23 === void 0 ? void 0 : _23.entry,
        webWorkerResolveAlias: (_24 = initialConfig.webWorker) === null || _24 === void 0 ? void 0 : _24.resolveAlias,
        webWorkerExternals: (_25 = initialConfig.webWorker) === null || _25 === void 0 ? void 0 : _25.externals,
        serverExternals: (_26 = initialConfig.server) === null || _26 === void 0 ? void 0 : _26.externals,
        webWorkerServerEntry: (_27 = initialConfig.webWorkerServer) === null || _27 === void 0 ? void 0 : _27.entry,
        jestConfig: jest,
        clientProjectName,
        clientFilesPath,
        isAngularProject: !!dependencies.angular || !!peerDependencies.angular,
        isReactProject: !!dependencies.react || !!peerDependencies.react,
        suricate: (_28 = initialConfig.suricate) !== null && _28 !== void 0 ? _28 : false,
        experimentalStorybook: (_29 = initialConfig.experimentalStorybook) !== null && _29 !== void 0 ? _29 : false,
    };
    return config;
};
