// yoshi-config/build/paths.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVER_BUNDLE = exports.SERVER_LOG_FILE = exports.STATS_FILE = exports.TSCONFIG_FILE = exports.SERVER_ENTRY = exports.POM_FILE = exports.NODE_PLATFORM_DEFAULT_CONFIGS_DIR = exports.ASSETS_DIR = exports.TEMPLATES_BUILD_DIR = exports.ROUTES_BUILD_DIR = exports.TYPES_DIR = exports.CJS_DIR = exports.ES_DIR = exports.STATICS_DIR = exports.PUBLIC_DIR = exports.TEMPLATES_DIR = exports.SERVER_CHUNKS_BUILD_DIR = exports.ROUTES_DIR = exports.DOCKER_FILE = exports.SERVER_CHUNKS_DIR = exports.TARGET_DIR = exports.BUILD_DIR = exports.SRC_DIR = void 0;
const path_1 = __importDefault(require("path"));
exports.SRC_DIR = 'src';
exports.BUILD_DIR = 'dist';
exports.TARGET_DIR = 'target';
exports.SERVER_CHUNKS_DIR = 'chunks';
exports.DOCKER_FILE = 'Dockerfile';
exports.ROUTES_DIR = path_1.default.join(exports.SRC_DIR, 'routes');
exports.SERVER_CHUNKS_BUILD_DIR = path_1.default.join(exports.BUILD_DIR, exports.SERVER_CHUNKS_DIR);
exports.TEMPLATES_DIR = path_1.default.join(exports.SRC_DIR, 'templates');
exports.PUBLIC_DIR = path_1.default.join(exports.SRC_DIR, 'assets');
exports.STATICS_DIR = path_1.default.join(exports.BUILD_DIR, 'statics');
exports.ES_DIR = path_1.default.join(exports.BUILD_DIR, 'esm');
exports.CJS_DIR = path_1.default.join(exports.BUILD_DIR, 'cjs');
exports.TYPES_DIR = path_1.default.join(exports.BUILD_DIR, 'types');
exports.ROUTES_BUILD_DIR = path_1.default.join(exports.BUILD_DIR, 'routes');
exports.TEMPLATES_BUILD_DIR = path_1.default.join(exports.STATICS_DIR, 'templates');
exports.ASSETS_DIR = path_1.default.join(exports.STATICS_DIR, 'assets');
exports.NODE_PLATFORM_DEFAULT_CONFIGS_DIR = 'test/configs';
exports.POM_FILE = 'pom.xml';
exports.SERVER_ENTRY = 'server';
exports.TSCONFIG_FILE = 'tsconfig.json';
exports.STATS_FILE = path_1.default.join(exports.TARGET_DIR, 'webpack-stats.json');
exports.SERVER_LOG_FILE = path_1.default.join(exports.TARGET_DIR, 'server.log');
exports.SERVER_BUNDLE = path_1.default.join(exports.BUILD_DIR, 'server.js');
