// yoshi-config/build/validConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jest_validate_1 = require("jest-validate");
const validConfig = {
    extends: 'preset',
    separateCss: jest_validate_1.multipleValidOptions(true, 'prod'),
    splitChunks: jest_validate_1.multipleValidOptions({}, false),
    cssModules: true,
    tpaStyle: false,
    enhancedTpaStyle: false,
    separateStylableCss: false,
    features: {
        externalizeRelativeLodash: false,
    },
    clientProjectName: 'project',
    keepFunctionNames: false,
    entry: jest_validate_1.multipleValidOptions('index.js', ['one.js'], { two: 'two.js' }, {
        app: 'index.js',
    }),
    servers: {
        cdn: {
            url: 'http://localhost:3200',
            port: 3200,
            dir: 'statics',
            ssl: false,
        },
        app: {
            port: 4000,
        },
    },
    resolveAlias: {
        hello: 'world',
    },
    externals: jest_validate_1.multipleValidOptions(['React'], { react: 'React' }),
    specs: {
        browser: 'test/**/*.spec.js',
        node: 'test/**/*.spec.js',
    },
    petriSpecs: {
        onlyForLoggedInUsers: false,
        scopes: ['me'],
    },
    transpileTests: true,
    externalUnprocessedModules: ['react'],
    exports: '[name]',
    hmr: jest_validate_1.multipleValidOptions(true, 'auto'),
    liveReload: false,
    performance: jest_validate_1.multipleValidOptions({}, false),
    hooks: {
        prelint: 'npm run stuff',
    },
    umdNamedDefine: false,
    projectType: 'app',
    experimentalBuildHtml: false,
    experimentalMonorepo: false,
    experimentalMinimalPRBuild: false,
    experimentalRtlCss: false,
    experimentalUseAssetRelocator: false,
    yoshiServer: false,
    startUrl: jest_validate_1.multipleValidOptions('http://localhost:3000', [
        'http://localhost:3000/hello',
        'http://localhost:3000/world',
    ]),
    webWorker: {
        entry: jest_validate_1.multipleValidOptions('index.js', ['one.js'], { two: 'two.js' }, {
            app: 'index.js',
        }),
        resolveAlias: {},
        externals: jest_validate_1.multipleValidOptions(['React'], { react: 'React' }),
    },
    server: {
        externals: jest_validate_1.multipleValidOptions(['React'], /react/, { react: 'React' }),
    },
    webWorkerServer: {
        entry: jest_validate_1.multipleValidOptions('index.js', ['one.js'], { two: 'two.js' }, {
            app: 'index.js',
        }),
    },
    suricate: false,
    experimentalStorybook: false,
};
exports.default = validConfig;
