// yoshi-common/build/html-rendering-data-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class default_1 {
    constructor({ outputFileName }) {
        this.outputFileName = outputFileName;
    }
    apply(compiler) {
        compiler.hooks.emit.tapPromise('HtmlRenderingDataPlugin', async (compilation) => {
            var _a;
            const entryNames = Array.from(compilation.entrypoints.keys());
            const jsonResult = JSON.stringify({
                entries: entryNames.reduce((acc, entryName) => {
                    const filesNames = compilation.entrypoints
                        .get(entryName)
                        .getFiles();
                    return Object.assign(Object.assign({}, acc), { [entryName]: filesNames });
                }, {}),
                publicPath: (_a = compiler.options.output) === null || _a === void 0 ? void 0 : _a.publicPath,
            });
            compilation.assets[this.outputFileName] = {
                source: () => jsonResult,
                size: () => jsonResult.length,
            };
        });
    }
}
exports.default = default_1;
