// yoshi-common/build/webpack-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const webpack_utils_1 = require("./webpack-utils");
const helpers_1 = require("./utils/helpers");
class WebpackManager {
    constructor() {
        this.namedConfigs = [];
    }
    addConfigs(name, configs) {
        this.namedConfigs.push({ name, configs: configs.filter(helpers_1.isTruthy) });
    }
    async run() {
        const webpackConfigs = this.namedConfigs.reduce((acc, { configs }) => {
            return [...acc, ...configs];
        }, []);
        const webpackStats = await webpack_utils_1.runWebpack(webpackConfigs);
        return {
            getAppData: (configName) => {
                let appConfigStartIndex = 0;
                let appConfigEndIndex = 0;
                // Webpack's configs and stats are being spread in a flat list
                // This function iterate over each app and locate the configs and stats
                // that assosiate with the requested app
                for (const entry of this.namedConfigs) {
                    const { name, configs } = entry;
                    if (configName === name) {
                        appConfigEndIndex = appConfigStartIndex + configs.length;
                        break;
                    }
                    appConfigStartIndex += configs.length;
                }
                const stats = webpackStats.stats.slice(appConfigStartIndex, appConfigEndIndex);
                const configs = webpackConfigs.slice(appConfigStartIndex, appConfigEndIndex);
                return { configs, stats };
            },
        };
    }
}
exports.default = WebpackManager;
