// yoshi-common/build/storybook/storybook.webpack.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const webpack_1 = require("webpack");
const webpack_config_1 = require("../webpack.config");
const createCustomStorybookWebpackConfig = ({ projectRoot, reporter, }) => (config) => {
    const yoshiWebpackConfig = webpack_config_1.createBaseWebpackConfig({
        configName: 'client',
        isDev: true,
        devServerUrl: '',
        target: 'web',
        useTypeScript: false,
        typeCheckTypeScript: false,
        name: 'storybook',
    });
    config.devServer = {
        colors: true,
        reporter,
    };
    if (config.module && yoshiWebpackConfig.module) {
        config.module.rules = yoshiWebpackConfig.module.rules;
    }
    if (config.plugins && yoshiWebpackConfig.plugins) {
        config.plugins = config.plugins.concat(yoshiWebpackConfig.plugins, new webpack_1.DefinePlugin({
            PROJECT_ROOT: JSON.stringify(projectRoot),
        }));
    }
    if (config.resolve && yoshiWebpackConfig.resolve) {
        config.resolve.extensions = yoshiWebpackConfig.resolve.extensions;
    }
    return config;
};
exports.default = createCustomStorybookWebpackConfig;
