// yoshi-common/build/storybook/storybook-process.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const events_1 = require("events");
const wait_port_1 = __importDefault(require("wait-port"));
const storybook_webpack_config_1 = __importDefault(require("./storybook.webpack.config"));
const storyBookConfigFolder = path_1.default.join(__dirname, '../..', '.storybook');
class StorybookProcess extends events_1.EventEmitter {
    constructor({ port }) {
        super();
        this.start = async () => {
            try {
                require.resolve('yoshi-storybook-dependencies/package.json');
            }
            catch (e) {
                throw new Error('Please install yoshi-storybook-dependencies in order to run storybook');
            }
            this.emit('compiling');
            // eslint-disable-next-line import/no-extraneous-dependencies
            const storybook = require(`@storybook/react/standalone`);
            const webpackConfig = storybook_webpack_config_1.default({
                projectRoot: path_1.default.join(process.cwd(), 'src'),
                reporter: this.reporter,
            });
            storybook({
                mode: 'dev',
                webpackConfig,
                port: this.port,
                ci: true,
                quiet: false,
                configDir: storyBookConfigFolder,
            });
            await wait_port_1.default({
                port: +this.port,
                output: 'silent',
                timeout: 20000,
            });
            this.emit('listening', this.port);
        };
        this.reporter = (middlewareOptions, options) => {
            const { state, stats } = options;
            state
                ? this.emit('finished-log', { stats, port: this.port })
                : this.emit('compiling');
        };
        this.port = port;
    }
    static create({ port = 9009 }) {
        return new StorybookProcess({ port });
    }
}
exports.default = StorybookProcess;
